module/zfs/dnode.c

```diff
@@ -1778,15 +1778,24 @@ dnode_try_claim(objset_t *os, uint64_t object, int slots)
 }
 
 /*
- * Checks if the dnode contains any uncommitted dirty records.
+ * Checks if the dnode itself is dirty, or is carrying any uncommitted records.
+ * Its important to check both, as some operations (eg appending to a file) can
+ * dirty both as a single logical unit, but they are not synced out atomically,
```

Comment:
dirty both --->
dirty both the dnode itself and any dnode range (resp. uncommitted block) (resp. uncommitted L0 block) (choose whichever is more appropriate) to be synced